extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/devui/DatabaseInspector.java

```diff
@@ -0,0 +1,174 @@
+                try (ResultSet tables = metaData.getTables(null, null, "%", new String[] { "TABLE" })) {
+                    while (tables.next()) {
+                        String tableName = tables.getString("TABLE_NAME");
+                        String tableSchema = tables.getString("TABLE_SCHEM");
+
+                        // Get the Primary Keys
+                        List<String> primaryKeyList = new ArrayList<>();
+                        try (ResultSet primaryKeys = metaData.getPrimaryKeys(null, tableSchema, tableName)) {
+                            while (primaryKeys.next()) {
+                                String primaryKeyColumn = primaryKeys.getString("COLUMN_NAME");
+                                primaryKeyList.add(primaryKeyColumn);
+                            }
+                        }
+
+                        // Get columns for each table
+                        List<Column> columnList = new ArrayList<>();
+                        try (ResultSet columns = metaData.getColumns(null, tableSchema, tableName, "%")) {
+                            while (columns.next()) {
+                                String columnName = columns.getString("COLUMN_NAME");
+                                String columnType = columns.getString("TYPE_NAME");
+                                int columnSize = columns.getInt("COLUMN_SIZE");
+                                String nullable = columns.getString("IS_NULLABLE");
+                                columnList.add(new Column(columnName, columnType, columnSize, nullable));
+                            }
+                        }
+                        tableList.add(new Table(tableSchema, tableName, primaryKeyList, columnList));
+                    }
+                }
+            }
+
+        } catch (SQLException ex) {
+            throw new RuntimeException(ex);
+        }
+
+        return tableList;
+    }
+
+    public DataSet executeSQL(String datasource, String sql) {
+        try {
+            AgroalDataSource ads = (AgroalDataSource) checkedDataSources.get(datasource);
+
+            try (Connection connection = ads.getConnection()) {
+                try (Statement statement = connection.createStatement()) {
+                    try (ResultSet resultSet = statement.executeQuery(sql)) {
+
+                        // Get the cols
+                        List<String> cols = new ArrayList<>();
+                        ResultSetMetaData metaData = resultSet.getMetaData();
+                        int columnCount = metaData.getColumnCount();
+                        for (int i = 1; i <= columnCount; i++) {
+                            String columnName = metaData.getColumnName(i);
+                            cols.add(columnName);
+                        }
+
+                        // Get the data
+                        List<Map<String, String>> rows = new ArrayList<>();
+                        while (resultSet.next() && rows.size() < 100) { // Linit to 100
```

Comment:
Note some JDBC drivers may fetch the whole result set in memory even if you only access the first few rows... this could lead to `OutOfMemoryError` if we're not careful.

Things like `java.sql.Statement#setFetchSize` can help -- though even that is just a hint, and IIRC the MySQL JDBC driveer will ignore it unless it's set to `Integer.MIN_VALUE`...

Reply:
yes, result set only filtering is not going to work for.

is scrollableResults honored by most drivers these days? maybe that could work?

Reply:
> is scrollableResults honored by most drivers these days? maybe that could work?

`ScrollableResults` is a Hibernate ORM feature. Not available using just JDBC.

As far as I know, with JDBC you can only use `ResultSet`, and like I said behavior varies depending on the JDBC driver.